### ISSUE TITLE
twisted.web2.auth.wrapper's interface is backwards

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2042]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2042|
|Type|enhancement|
|Created|2006-08-31 17:47:41Z|
|Branch|https://github.com/twisted/twisted/tree/web2cred-2042|

The correct way to do authentication with guard is for the returned object to be the resource which is rendered.  There's an example of this in nevow.guard; the code in Nevow may freely be moved into Twisted as far as it makes sense.

The major problem with not using the returned object is that there's no way to differentiate the authorization levels of different users, and application logic moves from the Realm (which is really where the majority of any Twisted application's logic should be kicked off) into the web-deployment code.

<details><summary>Searchable metadata</summary>

```
trac-id__2042 2042
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__Web2_Gold_Master Web2-Gold-Master
branch__web2cred_2042 web2cred-2042
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web2 web2
keywords__ 
time__1157046461000000 1157046461000000
changetime__1215791306000000 1215791306000000
version__None None
owner__ 
cc__therve cc__oubiwann cc__jknight cc__exarkun cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_2"></a>|@dreid commented|
|-|-|

I've been mulling this over, and I really need to fix this for webdav ACLs to not be a horribly convaluted mess that misuses cred.    The problem really hinges on the fact that nothing is ever done with the return value of portal.login()  This return value will be the application specific Avatar.  I propose a new interface that provides access to this object via the request.  Something along the following lines:

```
class IAuthenticatedRequest(Interface):
  """ A request that has been authenticated with the use of Cred and 
  holds a reference to the avatar return by portal.login.
  """
  avatar = Attribute("The application specific avatar returned by portal.login")
```

A second interface would also be of use, this interface would be used by web2 subsystems such as logging or CGI/SCGI to retrieve the username from the IAuthenticatedRequest.avatar.  Providing it should of course be optional.

```
class IHTTPUser(Interface):
  username = Attribute("A string representing the 'username' portion of the credentials used for authentication")

```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_3"></a>|@dreid commented|
|-|-|

branched to [branches/web2-auth-request-2042](branches/web2-auth-request-2042)  This branch covers fixes for [#1827](https://github.com/twisted/twisted/issues/1827) as well.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @dreid|
|-|-|

OK, some formatting comments:

 * unused import in wrapper
 * unused imports in test_httpauth
 * trailing whitespaces
 * IHTTPUser doesn't have a docstring
 * IAuthenticatedRequest should use new format for docstring (first line clear)

Regarding IHTTPUser, it's not used anywhere, should it be used in the tests ? You redefine ITestHTTPUser but I guess it would be better to use IHTTPUser.

Another problem is that doc/web2/examples/auth is broken, so I guess the change is important enough to break user programs (I think it comes from the change in locateChild). I've nothing against that, after all web2 is not stated as stable, but we should advertise it.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_6"></a>|@dreid removed owner|
|-|-|

Thanks for catching the web2 examples.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @dreid|
|-|-|

OK, IHTTPUser is not imported in test_httpauth. Interface and Attribute imports could be removed.

I'm puzzled about the whitespace changes in digest and basic, because we generally don't clean the file we don't touch.

Sorry for nit-picking.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid removed owner|
|-|-|

Missing import fixed.  As for the whitespace changes to digest and basic, i just cleaned all open files in my editor, if it really bothers you can I revert those but I don't think it's a big deal.

Thanks again for the review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @dreid|
|-|-|

It's almost perfect. If you can:
 * remove Interface and Attribute import from test_httpauth
 * new test methods should be named test_fooBar.

Please merge, and be sure your commit message notifies users and release manager about the change.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_10"></a>|@dreid set status to `closed`|
|-|-|

(In [19618]) Merge web2-auth-request-2042

Author: dreid
Reviewer: therve

Fixes [#2042](https://github.com/twisted/twisted/issues/2042)
Fixes [#1827](https://github.com/twisted/twisted/issues/1827)

Address deficiencies in the web2 HTTPAuthWrapper implementation.  
Add two new interfaces IAuthenticatedRequest, and IHTTPUser that
may be used to access the avatar for the authenticated user and 
get information about that user from the avatar.  

Also address two bugs in HTTPAuthWrapper that caused incorrect
reporting of allowedMethods and the first segment of the url to
be lost when returning the wrapped resource from locateChild.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `reopened`|
|-|-|

The interface is still backwards.  It might not be necessary to the revert the branch, but the issue described by this ticket is still outstanding.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

Hi.  This ticket was just brought to my attention again, and I apologize for the unclear description and the lack of reviews and stuff.

To clarify what I meant in the description of the ticket, ```IHTTPUser``` is not the cred interface  It's not clear to me that ```IHTTPUser``` should exist anywhere; it doesn't provide any value over the 'username' string.  The cred interface must be ```IResource``` to be semantically compatible with other cred implementations.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_14"></a>|@dreid commented|
|-|-|

I have yet to be convinced that I care about semantic compatibility with guard.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_15"></a>|@jyknight commented|
|-|-|

Some reasons why the current user needs to be accessible from resources down the tree:
 * Running CGI scripts, or things like CGI scripts (e.g. wsgi, scgi, etc.)
 * Putting the username in the logfile.
 * Serving a normal filesystem like thing via DAV.

These are reasons why the current method makes sense as the default. It is certainly possible to make a Resource which decides to use a completely separate subtree of resources for each user. That can make sense for some applications (let's say, for example, a webmail program). But it just doesn't seem to me that it fits the normal way that things work on the web. 

It doesn't make sense for anything I've wanted to do. I've always wanted to present the same resource tree to all users, and require authentication for a small subset of resources way down the tree. When I implemented this with nevow, I had to go to the extra effort of working _around_ all the stuff Nevow wanted to do for me.

I think the mismatch is that a website, even when everything lives under a single authentication realm, isn't a single application. It is a collection of applications which have possibly differing authentication needs. Thus, the point in the resource hierarchy where the authentication is established is often different than the place in the tree where you want to verify the authentication.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

The simplified API you're describing can be implemented in terms of the more conventional usage of cred.  Conversely, the API described on this ticket (which I suppose is what web2 more or less currently implements?) does not allow the conventional usage of cred.

If the API described here is so important (I don't think it is), then it should be offered as a utility layer on top of an implementation of what Glyph is describing.  It can be just as easy to use as if it is the "native" web2 authorization mechanism (and easier, since when someone has been using it and realizes they need something with more functionality, they will be able to switch to the underlying mechanism).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

If you look at Nevow, the factoring on this is basically correct.

If you want the username that is attached to the request, you ask the request.  That's an HTTP API that is encapsulated by the Request object.  That is _not_ the way that Twisted itself does authorization.

If you want to get a logged in user from a Twisted realm, you use Cred, and the cred interface is ```IResource```.  You might want to be able to have a CGI script pretend to have a _different_ HTTP username, depending on your cred realm.  That is also a separate problem, which, if you're logged in to an ```IResource```, can be trivially achieved by wrapping the request object in your realm resource.  If all you get is a string passed through, then you have to literally interpret what the web client is sending you, which is horribly inflexible AND incompatible with every single other protocol implementation within Twisted.

James, you're correct in saying that a website is (potentially) a collection of applications with different authentication needs.  That's why guard is a _resource_, rather than being a Site or a channel or something further up the protocol hierarchy.  You can be logged in to two realms on the same domain by placing two different guard instances at two different places in the resource hierarchy.   You can integrate applications by creating resources with the appropriate attributes and collecting them if the same user's credentials apply to multiple applications in different ways.  The use-case you described perfectly fits in, and you shouldn't have to work around anything to get it to happen (although granted, Nevow's guard implementation is kind of ad-hoc and grotty and there may well be issues I don't know about that applied in that particular case).

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_18"></a>|@jyknight commented|
|-|-|

Yes, I want an authenticated user, or an anonymous user, or whatever. I don't just want whatever the client passed me as a raw string. Transforming the client input into some user object is good. (that user object must have a username attached to it, but may have whatever other info is desired as well).

I want this user object, not the Basic Auth username, accessible from anywhere in the resource tree.

This is what web2 does right now. 

What I don't have any need for is the user represented as Resource. *That* is the part that just doesn't make sense to me.

> You can integrate applications by creating resources with the appropriate attributes and collecting them if the same user's credentials apply to multiple applications in different ways. The use-case you described perfectly fits in, and you shouldn't have to work around anything to get it to happen (although granted, Nevow's guard implementation is kind of ad-hoc and grotty and there may well be issues I don't know about that applied in that particular case).

All I wanted: a website-wide auth domain (thus, established at the top-level), under which exists the same resource tree for everyone. Some resources, though, care which user accesses them, and will either: deny the request (if logged in with a not-priviledged-enough account), request that the user login if they have a login (if currently anonymous), or present different data (if it's something like a "who am I" page). Some may even present a different sub-tree of resources depending on the user. Most resources, though, are completely agnostic. Some resources were actually implemented as external CGI programs (for historical reasons), but still want the same pre-authenticated user info.

This jives with the way web2's auth works. It doesn't appear to jive with the "correct" way to use cred.

I don't buy the claim that what's there right now provides provides reduced functionality; if you want to, you can dispatch to a different resource tree based on the user object, which as far as I understand is what I think you want to be able to do. And if you do that dispatch separately from the auth mechanism, you can do it anywhere in the tree.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Replying to [jknight](#note_18):
> Yes, I want an authenticated user, or an anonymous user, or whatever. I don't just want whatever the client passed me as a raw string. Transforming the client input into some user object is good. (that user object must have a username attached to it, but may have whatever other info is desired as well).
>
> I want this user object, not the Basic Auth username, accessible from anywhere in the resource tree.

Glad we at least all agree on this. :)

>
> This is what web2 does right now.
>
> What I don't have any need for is the user represented as Resource. *That* is the part that just doesn't make sense to me.

It's a superset of what you described above.  If the additional functionality doesn't make sense to you, can you at least see that it makes sense to other people, and so it should be provided, since it does not conflict with the use-case you have in mind?

>
> > You can integrate applications by creating resources with the appropriate attributes and collecting them if the same user's credentials apply to multiple applications in different ways. The use-case you described perfectly fits in, and you shouldn't have to work around anything to get it to happen (although granted, Nevow's guard implementation is kind of ad-hoc and grotty and there may well be issues I don't know about that applied in that particular case).
>
> All I wanted: a website-wide auth domain (thus, established at the top-level), under which exists the same resource tree for everyone. Some resources, though, care which user accesses them, and will either: deny the request (if logged in with a not-priviledged-enough account), request that the user login if they have a login (if currently anonymous), or present different data (if it's something like a "who am I" page). Some may even present a different sub-tree of resources depending on the user. Most resources, though, are completely agnostic. Some resources were actually implemented as external CGI programs (for historical reasons), but still want the same pre-authenticated user info.
>
> This jives with the way web2's auth works. It doesn't appear to jive with the "correct" way to use cred.

It's not a use-case I much care about, and it mostly ignores a big part of what makes cred great, but that doesn't mean it doesn't jive with cred.

Consider this:

```python
class YourRealm(object):
    implements(IRealm)

    def __init__(self, resource):
        self.resource = resource

    def requestAvatar(self, avatarId, mind, *interfaces):
        for iface in interfaces:
            if IResource.isEqualOrExtendedBy(iface):
                return iface, self.resource, lambda: None
```

Essentially, the cred-ly "correct" implementation of the behavior you want, given a web2 auth system which behaves in the cred-ly "correct" way.  If you want, replace ```resource``` with ```resourceFactory``` so you can have a different instance per-whatever.  Or make it take avatarId as an argument so it can do conditionals on it.  Or have the realm set it as an attribute.  etc etc etc.

Does this make any more sense?  Or did you already understand this was possible and am I still not understanding your objection?

>
> I don't buy the claim that what's there right now provides provides reduced functionality; if you want to, you can dispatch to a different resource tree based on the user object, which as far as I understand is what I think you want to be able to do. And if you do that dispatch separately from the auth mechanism, you can do it anywhere in the tree.
>

Of course.  I thought someone would bring this up.  One can re-implement cred inside a resource to work around the fact that cred isn't being used "correctly" in the web server itself.  This doesn't seem like a reasonable solution to me.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_20"></a>|@jyknight commented|
|-|-|

Replying to [exarkun](#note_19):
> > This is what web2 does right now.
> >
> > What I don't have any need for is the user represented as Resource. *That* is the part that just doesn't make sense to me.
>
> It's a superset of what you described above.  If the additional functionality doesn't make sense to you, can you at least see that it makes sense to other people, and so it should be provided, since it does not conflict with the use-case you have in mind?

No, because it seems to me that it makes the system less flexible, not more. (of course, you can completely ignore the inflexible part and do things your own way, which is what I've done in the past.)

> Essentially, the cred-ly "correct" implementation of the behavior you want, given a web2 auth system which behaves in the cred-ly "correct" way.  If you want, replace ```resource``` with ```resourceFactory``` so you can have a different instance per-whatever.  Or make it take avatarId as an argument so it can do conditionals on it.  Or have the realm set it as an attribute.  etc etc etc.
>
> Does this make any more sense?  Or did you already understand this was possible and am I still not understanding your objection?

I understood that's possible. That's what I ended up doing for nevow. But I can see no value in forcibly tying the dispatching of the resource tree to the authentication. If you want to tie a user to a resource tree, the following will do whether you want to do it at the same level as the authentication, or anywhere else in the tree. This code of course assumes your implementation of IHTTPUser also contains a resource attribute (YourRealm would put it there.)

```python
class DispatchResource:
  implements(IResource)
  def locateChild(self, request, segs):
    return request.avatar.resource, segs

  def renderHTTP(self, request):
    return request.avatar.resource
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

Twisted applications should have their main-point in a Realm object.  This allows for development of multi-protocol applications without distinct deployment and configuration code for each different protocol.  When you deploy a twisted cred application into a web2 container, the way you should do that is dropping in an authentication and authorization resource which wraps a portal.

I understand that in many cases it is convenient to have a global "username" variable.  Failing global-ness, it's sometimes useful to have that variable be a property of the request, or the session, or the "context".  Lots of web application frameworks always do things that way.  PHP comes to mind.

However, this style of application architecture is plain old bad design.  While it should be _possible_ with Twisted, it _should_ feel awkward and inconvenient, because it can only and always result in brittle, buggy code.  As we've painfully discovered again and again and again in Nevow, trying to cram things into an unspecified application context container is asking for trouble.

Your brief example is a great demonstration of this antipattern.  ```DispatchResource``` claims to implement ```IResource```, but it lies.  The "request" object does _not_ specify an "avatar" attribute, so ```renderHTTP``` violates its contract by expecting more of its parameters than the interface specifies.  Even assuming it does, there's no possible interface that could be required of "avatar" which would always have a "resource" attribute.  In other words, this code only works once it has been wrapped in a cred object that arbitrarily mangles the request object, and will otherwise break in an undocumented and unspecified way.  If this is the suggested way of using cred, authenticated and non-authenticated applications suddenly behave completely differently.

This is the reason that all cred objects specify only behavior described by the protocol that they represent; in HTTP's case, ```IResource```.  You can make the same argument about "multiple cooperating applications" for just about any protocol and ask for the user's name in the same way; in IMAP, it isn't uncommon for different mailboxes to actually represent different applications.  Nevertheless, you have to have an implementation of ```IAccount``` which integrates different ```IMailbox``` implementations accordingly, and cred does not supply an ```IIMAPUsername``` object just because sometimes application code might want to know your email address.  You must treat the protocol session as an interface that any arbitrary realm can implement, with no requirements on its deployment.

To be clear: this is not simply an opinion that I have about HTTP authentication, this the **single most important property** of any Twisted protocol implementation.  If you disagree and you aren't going to implement it this way, then I will delay the official release of web2 until someone else has done it.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_22"></a>|@radix commented|
|-|-|

>No, because it seems to me that it makes the system less flexible, not more.

Can you come up with an example of an application that would need to take advantage of whatever flexibility you have in mind, and would be hindered by the wrapping strategy?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph set owner to dialtone<br>@glyph set status to `new`|
|-|-|

In the meanwhile, perhaps dialtone should implement this since he hasn't chimed in to say he hates it yet :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph removed owner|
|-|-|

Different plan!  I implemented it myself.  It's ready for review in ```web2cred-2042```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_25"></a>|@jyknight commented|
|-|-|

This branch doesn't appear to address any of the issues raised. Please do not merge before addressing them.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

I'm not sure what issues you are referring to.  A number of vague preferences have been stated and implementation strategies proposed on this ticket, but no use-case driven requirements that I can see.

I'd be happy to provide sample implementations of any specific authentication requirements in terms of the revised authentication in the branch, but I think that the other issues are just that: other issues, which should have other tickets.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

Actually, reading my last comment it seems a bit vague.

There _are_ use-cases described on this ticket.  Specifically your bulleted list:

  * Running CGI scripts, or things like CGI scripts (e.g. wsgi, scgi, etc.)
  * Putting the username in the logfile.
  * Serving a normal filesystem like thing via DAV. 

However, none of these use-cases are currently implemented, and therefore none of these are broken by the change proposed in the branch.  Furthermore, I've implemented each these things (well, not literal DAV, but a filesystem-like resource) using an API exactly like the one now in the branch, so I'm quite sure that they can be implemented using it.

If you don't like the style of authentication required by Twisted protocols, feel free to file and/or reassign tickets representing all of those use-cases to me, and I can implement them after this branch is merged.  Hopefully that will illustrate that the use of "correct" cred qua HTTP is neither difficult nor tedious, but in any case neither you nor dreid will need to deal with the primary internal uses of it yourselves.

I would also be happy to provide utility functions to do some of the things that you are doing with web code outside of Twisted itself conveniently.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_29"></a>|@dreid set owner to @glyph|
|-|-|

When I was originally trying to address this issue I did not call the logout portion of the cred result.  Itamar raised the issue that logout should probably always be called (perhaps via a response filter?)  I'm honestly not really sure about it but thought I'd atleast bring it up for discussion.

Also the _loginSucceeded method now looks much denser than it is because you got rid of the whitespace around the _addAuthenticateHeaders closure.

Other than that I see nothing wrong with this change, since as you (glyph) said, nothing in twisted currently depends on the behaviour of the code/methodology it is replacing.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_30"></a>|@jyknight commented|
|-|-|

I'd still like to actually see at least some of the use cases addressed, since I currently don't see how they would be addressed in a way I'd like to use, wheras I can see directly how it's implemented without this change. 

Sample patches (e.g. without fulfilling normal doc/test requirements) would be fine with me. How does it work for log, and how doe it work for CGI.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_31"></a>|@glyph commented|
|-|-|

Replying to [jknight](#note_30):

> Sample patches (e.g. without fulfilling normal doc/test requirements) would be fine with me. How does it work for log, and how doe it work for CGI.

The more I think about this, the more I think that it ought to be a separate thing from this branch.

The issue here is that although HTTP itself doesn't have a notion of "username", various tools (CLF, CGI, SCGI) share a common hallucination that a request is associated with such a thing.  This is effectively the ad-hoc standardization of an HTTP daemon design antipattern, albeit an extremely subtle one.  Much as I don't _like_ this standardization, and much as it is not reflected in any actual spec, it does exist and we should cope with it elegantly.

So, it seems to me that the "request" object in web(2) ought to have such a string associated with it.  As an artifact parseable _into_ an HTTP request through various channels, (```REMOTE_USER``` environment var, for example, when web(2) is a CGI application, rather than server) it should have a public mutator, and as an artifact that can be serialized _from_ an HTTP request object, it should have a public accessor.

_However_, that string is, and should be presented as, an integration artifact to present to other systems.  It is not the way to integrate with a _native_ Twisted system; that's cred.

There are two junction points with cred, though, which is why this has gotten so tricky.  Cred has an avatarID, which looks suspiciously like a username.  After various transformations it might not look very much like the thing that came in to the basic auth or client cert DN or what have you, but it is nevertheless probably a reasonable default value (in the case where you're using cred) to set on the request object.

To illustrate why these aren't always necessarily the same, though, it also seems like a relatively reasonable thing to eventually have an ```IUsernamePassword```-style cred integration point with the _client_ half of CGI and friends.  It would be reasonable to consider it pre-authenticated with the ```REMOTE_USER``` as the username.  Now that same string is serving a subtly different role.  Although I seriously doubt that one would run a Twisted web server with a Twisted web application in a CGI container, much less with the same realm and checkers and then expect everything to line up, I can't anticipate the insane chutes-and-ladders setups of apache and Twisted and ligHTTPd that might one day run some real-life website.  The point is that it might be necessary in some configurations to specifically adjust the username string after it's arrived in your request from some source which provided it anyway.

In other words, I think the answer here is "we're both right".  The cred application interface _is_ ```IResource```, but the request should also have a simple username string to identify it to systems which expect that an HTTP resource has one.

(Then there's the issue of session identification and its implications for the 'logout' callback, but that is neither here nor there, I suppose.)

Ideally, I don't really think that these two requirements should be implemented together, but then, HTTP is hardly ideal.  I think I understand the concern you have now.

Rather than provide sample implementations of CGI and logging, I think I'm going to add the "reasonable default" behavior that I described to the cred implementation, and the username string to the request in this branch.  I can do more if it's still not obvious at that point how to muck around with CLF compliance or set and parse CGI environment variables.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_32"></a>|@glyph commented|
|-|-|

Replying to [glyph](#note_31):
> In other words, I think the answer here is "we're both right".  The cred application interface _is_ ```IResource```, but the request should also have a simple username string to identify it to systems which expect that an HTTP resource has one.

Gyah.  "which expect that an HTTP _request_ has one."

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_33"></a>|@jyknight commented|
|-|-|

I think it sounds like we're on the same page now. I agree that the current code is using cred wrong, and I'm glad you've come to agree that the request should have a pre-authenticated username string stored on it by cred. That makes it again obvious to me how to implement logging/CGI, so I will retract my request for an implementation. Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_35"></a>|@dreid commented|
|-|-|

A brief unsolicited review.  This looks almost completely right.  The only thing it seems to be lacking is the pre-authenticated username string stored on the request.  My initial reaction is that this task belongs to the credential factories, since getting the pre-authenticated username string is going to be specific to the type of authentication being used.  For Digest and Basic it's the username key, for Negotiate it's harder to get at and requires calling out to the Kerberos APIs.

I'm not entirely sure but I think the _addAuthenticateHeaders responseFilter is unecessary.  If I recall it was there so that later resources could force authentication.  I think exarkun is right about the correct thing to do being to add another HTTPAuthResource wrapper later in the tree.

Also (slightly outside the scope of this ticket) I'm wondering how best to integrate this improved HTTPAuthResource with a form+cookie based mechanism.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_36"></a>|@glyph set status to `assigned`|
|-|-|

dreid, thanks for moving the ball on this.  I'll have to consider your comments later, but I guess this isn't actually too far away from done, and I should just finish it up soon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun commented|
|-|-|

Suggesting close as wontfix.  The web2 HTTP auth code has been ported to `twisted.web` and presents the usual guard interface.  The auth stuff in web2 should just be dropped whenever `twisted.web` is good enough that we don't need to have `twisted.web2` around at all anymore.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
